INT-3387: MessageGroupStore improvements

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3387,
https://jira.spring.io/browse/INT-3806
- Introduce

```
MessageGroupStore

void addMessagesToGroup(Object groupId, Message<?>... messages);
```

And implement it in all stores.
- Use new `addMessagesToGroup` where it is reasonable, e.g. `DelayHandler`
- Optimize test-case to use a new store method (where it is possible)
- Fix timing delays in the `JdbcMessageStoreTests`
- Introduce `PersistentMessageGroup`
- Add `AbstractMessageGroupStore#proxyMessageGroupForLazyLoad` to wrap the raw `MessageGroup` to the `PersistentMessageGroup` for lazy-load
- Rework `MessageGroupMetadata` do not be `immutable` and allow to store/restore in the `AbstractKeyValueMessageStore` only the `MessageGroupMetadata`
- Refactor `ResequencingMessageHandler` and `SequenceSizeReleaseStrategy` a bit for better performance when interact with the `MessageGroup`
- Add `AbstractMessageGroupStore#setLazyLoadMessageGroups` to switch off the `lazy-load` behavior and restore the previous full `MessageGroup` logic
- Add `What's New` note and `message-store.adoc` paragraph for the lazy-load functionality

artembilan committed:
INT-3387: MessageGroupStore improvements

JIRA: https://jira.spring.io/browse/INT-3387,
https://jira.spring.io/browse/INT-3806

* Introduce
```
MessageGroupStore

void addMessagesToGroup(Object groupId, Message<?>... messages);
```
And implement it in all stores.

* Use new `addMessagesToGroup` where it is reasonable, e.g. `DelayHandler`
* Optimize test-case to use a new store method (where it is possible)
* Fix timing delays in the `JdbcMessageStoreTests`
* Introduce `PersistentMessageGroup`
* Add `AbstractMessageGroupStore#proxyMessageGroupForLazyLoad` to wrap the raw `MessageGroup` to the `PersistentMessageGroup` for lazy-load
* Rework `MessageGroupMetadata` do not be `immutable` and allow to store/restore in the `AbstractKeyValueMessageStore` only the `MessageGroupMetadata`
* Refactor `ResequencingMessageHandler` and `SequenceSizeReleaseStrategy` a bit for better performance when interact with the `MessageGroup`
* Add `AbstractMessageGroupStore#setLazyLoadMessageGroups` to switch off the `lazy-load` behavior and restore the previous full `MessageGroup` logic
* Add `What's New` note and `message-store.adoc` paragraph for the lazy-load functionality

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/PersistentMessageGroup.java:
```diff
@@ -0,0 +1,224 @@
+		else {
+			Message<?> message = getOne();
+			if (message != null) {
+				return message.getHeaders().get(IntegrationMessageHeaderAccessor.SEQUENCE_SIZE, Integer.class);
```
NPE if header missing.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/MessageGroupStore.java:
```diff
@@ -140,14 +129,31 @@
+	 * @param messages The messages to add.
+	 * @since 4.3
+	 */
+	void addMessagesToGroup(Object groupId, Message<?>... messages);
```
What is the purpose of this? I don't see it called with more than one message and it adds overhead in that case.

artembilan replied:
That's awry here because we can't have an overloaded method with different return type like an existing already. See `BasicMessageGroupStore`:
```
MessageGroup addMessageToGroup(Object groupId, Message<?> message);
```
I did like this to align with the:
```
void removeMessagesFromGroup(Object key, Collection<Message<?>> messages);
```

But I'm open for any other idea.

I'm really not going to deprecate `addMessageToGroup()`, BTW.

garyrussell commented:
I need to dig deeper, but this looks good; I am not so sure lazy should be the default, it's probably worse for small groups. I wonder if we could switch if, say, the group size exceeds, say, 5 messages.

artembilan committed:
`GroupType.PERSISTENT` and not lazy by default

artembilan commented:
Pushed.
Now it isn't `lazy` by default.

> I wonder if we could switch if, say, the group size exceeds, say, 5 messages.

Does not make sense. We load the whole group only once in any case. But in many aggregator cases, without lazy, we do that for nothing. So, even if it is enough small group, loading it fully to the memory without a practical usage is a some overhead anyway.

garyrussell commented:
My point is that one select returning 5 messages is much faster than 5 selects for one message at a time.

artembilan commented:
> My point is that one select returning 5 messages is much faster than 5 selects for one message at a time.

??? Where do you see such a logic?
`Lazy-Load` means: do load the messages on demand, not one by one

garyrussell commented:
I might have exaggerated - :smile: - I am thinking of cases where we `findOne()`, decide to release the group and then `getMessages()` - 2 roundtrips instead of one.

artembilan commented:
> I might have exaggerated

M-m-m. Should I spend some time for the benchmarks to be sure that we really need that extra logic for the threshold? 
Right now I only have visual observation from the `AbstractCorrelatingMessageHandler`:
````java
MessageGroup messageGroup = this.messageStore.getMessageGroup(correlationKey);
...
if (!messageGroup.isComplete() && messageGroup.canAdd(message)) {
...
	messageGroup = this.store(correlationKey, message);
	if (this.releaseStrategy.canRelease(messageGroup)) {
...
			completedMessages = this.completeGroup(message, correlationKey, messageGroup);
````
So, with lazy-load we have the whole group only after the `canRelease()` in the `completeGroup()`.

With the `threshold` variant we load the whole group on each `getMessageGroup` and/or `store()` for each incoming message.

Am I missing anything?

garyrussell commented:
I don't think it's worth it, I am probably just paranoid; I don't mind lazy being the default as long as there is clear documentation about the trade-off and how to switch between the two. Users can do their own benchmarks since only they have real data to compare.

artembilan committed:
PR Comments

artembilan commented:
Restore the `lazy-load` as a default behavior.

artembilan committed:
Fix `JdbcMessageStoreTests` timing issues

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java:
```diff
@@ -94,38 +94,62 @@ public long getMessageCount() {
+		MessageGroupMetadata metadata = getGroupMetadata(groupId);
+		SimpleMessageGroup group = null;
+		if (metadata == null) {
+			group = new SimpleMessageGroup(groupId);
```
Shouldn't we always use the group factory?

This ctor creates a group with a list when a set (default for the factory) is more efficient for removes.

artembilan replied:
This case doesn't matter.
We don't remove from the group, neither return this object from the `addMessagesToGroup` method.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/MessageGroupFactory.java:
```diff
@@ -36,4 +36,10 @@
+
+	MessageGroup create(MessageGroupStore messageGroupStore, Object groupId, long timestamp,
+			boolean complete);
+
```
We need javadocs on all methods in this interface.

garyrussell commented:
Looks good; needs a rebase and I'd like to see some performance tests, say with aggregating 1000 messages and a JDBC message store; we should then indicate the improvement in the docs; it might even warrant a blog post.

artembilan committed:
Address PR comments

* Add performance test to the `ConfigurableMongoDbMessageGroupStoreTests`
* Add JavaDocs for the `MessageGroupFactory` methods
* Add `log4j.properties` into the `test` MongoDB module for better traceability
* Fix `JdbcMessageStore#getOneMessageFromGroup()` over the `doPollForMessage()` delegation.
The `jdbcTemplate.queryForObject()` requires exactly one and only one raw in `resultSet`
* Add performance test results into the `message-store.adoc`

artembilan commented:
Pushed.

artembilan commented:
Not related to this PR, but one of your latest changes:
```
org.springframework.integration.amqp.outbound.AsyncAmqpGatewayTests > testConfirmsAndReturns FAILED

    java.lang.AssertionError at AsyncAmqpGatewayTests.java:176
```
So, I wonder what may cause the `ack` not being delivered.
Works well locally though...

garyrussell commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/store/ConfigurableMongoDbMessageGroupStoreTests.java:
```diff
@@ -72,6 +76,48 @@ public void testWithAggregatorWithShutdown() throws Exception {
+
+		performLazyLoadEagerTest(watch, sequenceSize, false);
+
+		System.out.println(watch.prettyPrint());
```
This will fail checkstyle `//    System.  out.  println` :smile:

artembilan replied:
Doh... Of course!
Feel free to fix on merge :wink: 
Thanks

garyrussell commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/store/mongo-aggregator-confugurable-config.xml:
```diff
@@ -6,7 +6,7 @@
 		http://www.springframework.org/schema/integration http://www.springframework.org/schema/integration/spring-integration.xsd">
 
 	<int:aggregator input-channel="inputChannel" output-channel="outputChannel" message-store="mongoStore"
```
typo (not new) `confugurable`.

garyrussell commented on src/reference/asciidoc/message-store.adoc:
```diff
@@ -106,3 +106,28 @@ when it would be an overhead to load entire `MessageGroup` from the store on eac
+02652  007%  Lazy-Load
+36266  093%  Eager
+....
+
```
Cool - I'll fix the typo on merge.

garyrussell commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/store/mongo-aggregator-confugurable-config.xml:
```diff
@@ -6,7 +6,7 @@
 
 	<int:aggregator input-channel="inputChannel" output-channel="outputChannel" message-store="mongoStore"
-					release-strategy-expression="size() == 10"/>
+					release-strategy-expression="size() == 1000"/>
```
I had to revert this - will add a note to the `@Ignored` test.

garyrussell commented:
Merged as 286c421c1a57dbda9cbf92f003e20bf4e0fec715
